File: 24.swap_nodes_in_pairs/swapPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if(head === null || head.next === null) return head;
    var dummy = new ListNode(-1);
    dummy.next = head;
    var former = dummy;
    var first = head;
    var second = head.next;
    
    do {//use do sentence can solve the problem that there are only 2 elements to be swapped
         first.next = second.next;
        second.next = first;
        former.next = second;
        
        if(first.next !==null&&first.next.next !==null){//verify next element first and then the next
        former = first;
        first = former.next;
        second = first.next;
        }else{
           return dummy.next;
        }
    }
    
    while (second !== null);
    return dummy.next;
};

File: 24.swap_nodes_in_pairs/swapPairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        former = dummy
        while head and head.next:
            #record the next node
            nextnode = head.next.next
            #swap nodes, start from former
            former.next = head.next
            head.next.next = head
            head.next = nextnode

            former = former.next.next
            head = nextnode
        # no head.next
        if head:
            former.next = head
        return dummy.next
